File: codechef/2012_August_Challenge/HOMDEL/HomeDelivery.cxx
#include <cstdio>
#include <algorithm>
using namespace std;
typedef long long LL;

// Floyd warshall 
int main(int argc, char *argv[]) {
	int N; scanf("%d", &N);
    LL T[250][250];
	for(int row = 0; row < N; ++row)
		for(int col = 0; col < N; ++col)
			scanf(" %lld", &T[row][col]);
    // find the shortest between each pair.
	for(int i = 0; i < N; ++i)
		for(int j = 0; j < N; ++j)
			for(int k = 0; k < N; ++k)
                T[i][j] = min(T[i][j], T[i][k]+T[k][j]);
    int M; scanf(" %d", &M);
	for(int scenario = 0; scenario < M; ++scenario) {
		int S,G,D;
		scanf(" %d %d %d", &S, &G, &D);
		
        LL timeTaken = T[S][G] + T[G][D];
        LL timeWasted = max(0ll, timeTaken-T[S][D]);
		printf("%lld %lld\n", timeTaken, timeWasted);
	}
	return 0;
}

File: CodeChef/2012_August_Challenge/HOMDEL/HomeDelivery.py
# Floyd warshall 
def main():
    N = int(raw_input())
    T = []
    for row in xrange(N):
        T.append([int(x) for x in raw_input().split()])
    # find the shortest between each pair.
    for i in xrange(N):
        for j in xrange(N):
            for k in xrange(N):
                T[i][j] = min(T[i][j], T[i][k]+T[k][j])
    M = int(raw_input())
    for scenario in xrange(M):
        S, G, D = tuple([int(x) for x in raw_input().split()])
        timeTaken = T[S][G] + T[G][D]
        timeWasted = max(0, timeTaken-T[S][D])
        print timeTaken, timeWasted

if __name__ == "__main__":
    main()
